File: Tests/LatticeTests/SyncTests/SyncMeasurementLogging.swift
import Foundation
@testable import Lattice

/// Explicit setup for the selected visibility fixture only. This does not
/// alter BaseTest or production defaults. The public nil setter selects stderr;
/// there is no public sink getter, so the receipt does not claim sink readback.
enum SyncMeasurementLogging {
    static let policy = "lattice.visibility.logging/1;native=off;native-sink=stderr;observer-worker=off;ack-path=off;sql-dump=absent;swift-log-env=absent"

    static func installIfRequested() throws -> [String: String]? {
        let env = ProcessInfo.processInfo.environment
        // Ordinary suite invocations retain their original process logger.
        guard let requested = env["LATTICE_SYNC_VISIBILITY_LOGGING_POLICY"] else { return nil }
        guard requested == policy else { throw SyncVisibilityFailure.invalid("unknown measurement logging policy") }
        for key in ["LATTICE_DUMP_SQL", "LOG_LEVEL"] {
            guard env[key] == nil else { throw SyncVisibilityFailure.invalid("measurement requires absent \(key)") }
        }
        for key in ["LATTICE_ACK_PATH_DIAGNOSTICS", "LATTICE_OBSERVER_WORKER_DIAGNOSTICS"] {
            guard env[key] == nil || env[key] == "0" else {
                throw SyncVisibilityFailure.invalid("measurement requires disabled \(key)")
            }
        }
        Lattice.setLogLevel(.off)
        Lattice.setLogFile(nil)
        let observed = lattice_get_log_level().rawValue
        guard observed == 0 else { throw SyncVisibilityFailure.invalid("native logging off setter did not take effect") }
        return ["LOGGING": policy,
            "nativeLoggingLevelAtStart": String(observed),
            "nativeLoggingSink": "stderr; nil FILE setter returned; no sink getter",
            "nativeLoggingVerification": "level readback at start/end; not continuous monitoring",
            "diagnosticControls": "observer-worker=0;ack-path=0;sql-dump=absent",
            "swiftLogging": "LOG_LEVEL absent; pinned library defaults retained"]
    }

    static func finish(metadata: inout [String: String]) throws {
        let observed = lattice_get_log_level().rawValue
        metadata["nativeLoggingLevelAtEnd"] = String(observed)
        guard observed == 0 else { throw SyncVisibilityFailure.invalid("native logging level changed during visibility fixture") }
    }
}

File: Scripts/sync_measurement_logging.py
"""Measurement-only logging controls; no native/SQLite imports or execution."""
POLICY = "lattice.visibility.logging/1;native=off;native-sink=stderr;observer-worker=off;ack-path=off;sql-dump=absent;swift-log-env=absent"
CONTROL = 'LATTICE_SYNC_VISIBILITY_LOGGING_POLICY'
ATTESTATION = {
    'LOGGING': POLICY,
    'nativeLoggingLevelAtStart': '0',
    'nativeLoggingLevelAtEnd': '0',
    'nativeLoggingSink': 'stderr; nil FILE setter returned; no sink getter',
    'nativeLoggingVerification': 'level readback at start/end; not continuous monitoring',
    'diagnosticControls': 'observer-worker=0;ack-path=0;sql-dump=absent',
    'swiftLogging': 'LOG_LEVEL absent; pinned library defaults retained',
}


def apply_environment(original):
    """Return a controlled child environment without mutating the caller."""
    env = dict(original)
    # SQL dumping checks getenv presence, so even value '0' enables it.
    env.pop('LATTICE_DUMP_SQL', None)
    env.pop('LOG_LEVEL', None)
    env[CONTROL] = POLICY
    env.update(LATTICE_LOG_LEVEL='0', LATTICE_ACK_PATH_DIAGNOSTICS='0',
               LATTICE_OBSERVER_WORKER_DIAGNOSTICS='0')
    return env


def require_controls(env):
    for key in ('LATTICE_DUMP_SQL', 'LOG_LEVEL'):
        if key in env:
            raise ValueError('measurement requires absent ' + key)
    for key in ('LATTICE_ACK_PATH_DIAGNOSTICS', 'LATTICE_OBSERVER_WORKER_DIAGNOSTICS'):
        if env.get(key) not in (None, '0'):
            raise ValueError('measurement requires disabled ' + key)


def metadata_valid(metadata):
    return isinstance(metadata, dict) and all(metadata.get(k) == v for k, v in ATTESTATION.items())


def require_metadata(metadata):
    if not metadata_valid(metadata):
        raise ValueError('measurement logging setup/readback attestation missing or contradictory')
